lib/internal/readline/interface.js

```diff
@@ -789,10 +809,41 @@ class Interface extends InterfaceConstructor {
 
   [kLine]() {
     const line = this[kAddHistory]();
+    this.undoStack = [];
+    this.redoStack = [];
     this.clearLine();
     this[kOnLine](line);
   }
 
+  [kPushToUndoStack](text, cursor) {
+    this.undoStack.push({ text, cursor });
+    if (this.undoStack.length > kMaxUndoRedoStackSize) {
+      this.undoStack.shift();
+    }
+  }
+
+  [kUndo]() {
+    const entry = this.undoStack.pop();
+    if (!entry) return;
+
+    this.line = entry.text;
+    this.cursor = entry.cursor;
+
+    this.redoStack.push(entry);
+    this[kRefreshLine]();
+  }
+
+  [kRedo]() {
+    const entry = this.redoStack.pop();
+    if (!entry) return;
```

Comment:
If think we should check the length before attempting to `pop`.
```suggestion
    if (this.redoStack.length === 0) return;
    const entry = ArrayPrototypePop(this.redoStack);
```

Reply:
Same as the previous review.